mettle/worker/worker.go

```diff
@@ -549,6 +545,52 @@ func (w *worker) execute(action *pb.Action, command *pb.Command) *pb.ExecuteResp
 	}
 }
 
+// runCommand runs a command with a timeout, terminating it in a sensible manner.
+// Some care is needed here due to horrible interactions between process termination and
+// having an in-process stdout / stderr.
+func (w *worker) runCommand(cmd *exec.Cmd, timeout time.Duration) error {
+	w.stdout.Reset()
+	w.stderr.Reset()
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+	defer stdout.Close()
+	go func() {
+		io.Copy(&w.stdout, stdout)
+	}()
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return err
+	}
+	defer stderr.Close()
+	go func() {
+		io.Copy(&w.stderr, stderr)
+	}()
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	ch := make(chan error)
+	go func() {
+		ch <- cmd.Wait()
+	}()
+	select {
+	case err := <-ch:
+		return err
+	case <-time.After(timeout):
+		log.Warning("Terminating process for %s due to timeout", w.actionDigest.Hash)
+		cmd.Process.Signal(os.Signal(syscall.SIGTERM))
```

Comment:
you could probably use the command context.Done() for this if you were so inclined.

Reply:
I'm not creating it with CommandContext any more; bit :scared_hamster: of that one after all this.